### PR TITLE
Townyupdate

### PR DESCRIPTION
**Pull request by Fortifier42**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -279,16 +279,16 @@
         <dependency>
             <groupId>com.palmergames</groupId>
             <artifactId>Towny</artifactId>
-            <version>0.96</version>
+            <version>0.96.2.0</version>
             <scope>system</scope>
-            <systemPath>${basedir}/lib/Towny.jar</systemPath>
+            <systemPath>${basedir}/lib/Towny-0.96.2.0.jar</systemPath>
```
Comment (mcmonkey4eva) [inline pom.xml:284]: Fix the file names

**Comment by Fortifier42** [comment]
This should be updated now :)

File: src/main/java/com/denizenscript/depenizen/bukkit/bridges/TownyBridge.java [hunk 2]
```diff
@@ -1,148 +1,118 @@
...
+                if (NationTag.matches(attribute.getContext(1))) {
+                    return attribute.contextAsType(1, NationTag.class);
+                }
+                else {
+                    Debug.echoError("Unable to match: '" + attribute.getContext(1) + "' to a valid nation!");
+                    return null;
+                }
             }
-        }
+            Debug.echoError("Must specify a nation for tag <nation[<name>]>!");
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/bridges/TownyBridge.java:113]: `attribute.echoError`

File: src/main/java/com/denizenscript/depenizen/bukkit/events/towny/PlayerCreatesTownyTownScriptEvent.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    public PreNewTownEvent event;
+
+    @Override
+    public boolean couldMatch(ScriptPath path) {
+        return path.eventLower.startsWith("towny player creates town");
+    }
+
+    @Override
+    public boolean matches(ScriptPath path) {
+        return super.matches(path);
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/events/towny/PlayerCreatesTownyTownScriptEvent.java:48]: If no matching is done, leave the method off entirely

File: src/main/java/com/denizenscript/depenizen/bukkit/events/towny/PlayerCreatesTownyTownScriptEvent.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    // <--[event]
+    // @Events
+    // towny player creates town
+    //
+    // @Regex ^on towny player creates town$
+    //
+    // @Triggers when a player tries to create a new town. This fires just before the town is created allowing you to cancel the event.
+    //
+    // @Context
+    // <context.town_name> Returns the name of the town.
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/events/towny/PlayerCreatesTownyTownScriptEvent.java:22]: `cancel_message` context and determine should be documented

File: src/main/java/com/denizenscript/depenizen/bukkit/events/towny/TownyTownCreatedScriptEvent.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+import com.denizenscript.depenizen.bukkit.objects.towny.TownTag;
+import com.palmergames.bukkit.towny.event.NewTownEvent;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.Listener;
+
+public class TownyTownCreatedScriptEvent extends BukkitScriptEvent implements Listener {
+
+    // <--[event]
+    // @Events
+    // towny player creates town
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/events/towny/TownyTownCreatedScriptEvent.java:14]: This meta seems to have been copy/pasted from the other event without changes

File: src/main/java/com/denizenscript/depenizen/bukkit/objects/towny/NationTag.java [hunk 6]
```diff
@@ -164,45 +184,40 @@ public String getAttribute(Attribute attribute) {
...
+        registerTag("balance", (attribute, object) -> {
             try {
-                return new ElementTag(nation.getAccount().getHoldingBalance()).getAttribute(attribute.fulfill(1));
+                return new ElementTag(object.nation.getAccount().getHoldingBalance());
             }
             catch (EconomyException e) {
-                if (!attribute.hasAlternative()) {
-                    Debug.echoError("Invalid economy response!");
-                }
+                Debug.echoError("Invalid economy response!");
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/objects/towny/NationTag.java:192]: `attribute.echoError`

File: src/main/java/com/denizenscript/depenizen/bukkit/properties/towny/TownyChatPlayerProperties.java [hunk 7]
```diff
@@ -54,16 +53,8 @@ private TownyChatPlayerProperties(PlayerTag player) {
...
-        if (attribute == null) {
-            return null;
-        }
-
-        if (attribute.startsWith("townychat")) {
-            Chat plugin = (Chat) TownyChatBridge.instance.plugin;
-            ChannelsHolder holder = plugin.getChannelsHandler();
-            attribute = attribute.fulfill(1);
+    public static void registerTags() {
+        PropertyParser.<TownyChatPlayerProperties>registerTag("townychat", (attribute, object) -> {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/properties/towny/TownyChatPlayerProperties.java:57]: This wasn't tested, was it?
None of these tags will work.

Comment (Fortifier42) [reply src/main/java/com/denizenscript/depenizen/bukkit/properties/towny/TownyChatPlayerProperties.java:57]: Nothing Towny wise was tested yet as I hadn't finished it haha

File: src/main/java/com/denizenscript/depenizen/bukkit/events/towny/PlayerCreatesTownyTownScriptEvent.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    // <context.cancel_message> Returns the cancel message if the town creation was cancelled.
+    //
+    // @Determine
+    // cancel_message:ElementTag to determine a new cancel message.
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/events/towny/PlayerCreatesTownyTownScriptEvent.java:26]: This is formatted wrong

**Pull request closed without merging**